### PR TITLE
LLC Split Commit Part 1

### PR DESCRIPTION
**Pull request by jenniferdai**
* Allows servers to send split commit requests to controllers that recognize the parameter

### CODE REVIEW

File: pinot-common/src/main/java/com/linkedin/pinot/common/protocols/SegmentCompletionProtocol.java [hunk 1]
```diff
@@ -170,6 +170,7 @@ public String getUrl(String hostPort) {
       private long _buildTimeMillis;
       private long _waitTimeMillis;
       private int _extraTimeSec;
+      private boolean _splitEnabled;
```
Comment (mcvsubbu) [inline pinot-common/src/main/java/com/linkedin/pinot/common/protocols/SegmentCompletionProtocol.java]: split enbaled is not a request param

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/LLCSegmentCommitStart.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+
+public class LLCSegmentCommitStart extends ServerResource {
+  private static Logger LOGGER = LoggerFactory.getLogger(LLCSegmentCommitStart.class);
+  long _offset;
+  String _segmentNameStr;
+  String _instanceId;
+
+  @HttpVerb("get")
+  @Description("Starts an LLC segment commit")
```
Comment (mcvsubbu) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/LLCSegmentCommitStart.java]: some note about split commit

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/LLCSegmentCommitStart.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    if (response.equals(SegmentCompletionProtocol.RESP_COMMIT_CONTINUE)) {
+      LOGGER.info("Success, continuing segment commit. Response {}", response);
+      return new StringRepresentation(response.getStatus().toString());
+    } else {
+      LOGGER.info("Failure, response {}", response);
+      return new StringRepresentation(response.getStatus().toString());
+    }
+  }
+
+  boolean extractParams() {
```
Comment (mcvsubbu) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/LLCSegmentCommitStart.java]: base class the two of them

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/LLCSegmentCommitUpload.java [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+  String _segmentNameStr;
+  String _instanceId;
+
+  public LLCSegmentCommitUpload()
+      throws IOException {
+  }
+
+  @Override
+  @HttpVerb("post")
+  @Description("Uploads an LLC segment coming in from a server")
```
Comment (mcvsubbu) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/LLCSegmentCommitUpload.java]: note for split commit

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/LLCSegmentCommitUpload.java [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+  @Summary("Uploads an LLC segment coming in from a server")
+  @Paths({"/" + SegmentCompletionProtocol.MSG_TYPE_COMMIT})
```
Comment (mcvsubbu) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/LLCSegmentCommitUpload.java]: need another message type

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/LLCSegmentCommitUpload.java [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+  public Representation post(Representation entity) {
+    if (!extractParams()) {
+      return new StringRepresentation(SegmentCompletionProtocol.RESP_FAILED.toJsonString());
+    }
+    LOGGER.info("segment={} offset={} instance={} ", _segmentNameStr, _offset, _instanceId);
+
+    boolean success = uploadSegment(_instanceId, _segmentNameStr);
+    if (success) {
+      LOGGER.info("Uploded segment successfully");
+      return new StringRepresentation(SegmentCompletionProtocol.ControllerResponseStatus.COMMIT_SUCCESS.toString());
```
Comment (mcvsubbu) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/LLCSegmentCommitUpload.java]: return UPLOAD_SUCCESS

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/LLCSegmentCommitUpload.java [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+      if (dataFile == null) {
+        LOGGER.error("Segment not included in request. Instance {}, segment {}", instanceId, segmentNameStr);
+        return false;
+      }
+      // We will not check for quota here. Instead, committed segments will count towards the quota of a
+      // table
+      LLCSegmentName segmentName = new LLCSegmentName(segmentNameStr);
+      final String rawTableName = segmentName.getTableName();
+      final File tableDir = new File(baseDataDir, rawTableName);
+      final File segmentFile = new File(tableDir, segmentNameStr + _offset + _instanceId);
```
Comment (mcvsubbu) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/LLCSegmentCommitUpload.java]: implement a method to create the file name. For now, put it in the SegmentCompletionProtocol file. with a TODO to find a place for it

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/LLCSegmentCommitUpload.java [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+
+      // always delte if there
```
Comment (mcvsubbu) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/LLCSegmentCommitUpload.java]: spell

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/SegmentCompletionManager.java [hunk 9]
```diff
@@ -181,7 +181,7 @@ private synchronized SegmentCompletionFSM lookupOrCreateFsm(final LLCSegmentName
    * Otherwise, this method will return a protocol response to be returned to the client right away (without saving the
    * incoming segment).
    */
-  public SegmentCompletionProtocol.Response segmentCommitStart(final SegmentCompletionProtocol.Request.Params reqParams) {
+  public SegmentCompletionProtocol.Response segmentCommitStart(final SegmentCompletionProtocol.Request.Params reqParams, boolean splitEnabled) {
```
Comment (mcvsubbu) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/SegmentCompletionManager.java]: s/splitEnabled/isSplitCommit

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/LLCSegmentCommitEnd.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+  @Paths({"/" + SegmentCompletionProtocol.MSG_TYPE_COMMIT_END})
+  public Representation get() {
+    if (!extractParams()) {
+      return new StringRepresentation(SegmentCompletionProtocol.RESP_FAILED.toJsonString());
+    }
+    LOGGER.info("segment={} offset={} instance={} ", _segmentNameStr, _offset, _instanceId);
+
+    final SegmentCompletionProtocol.Request.Params reqParams = new SegmentCompletionProtocol.Request.Params();
+    reqParams.withInstanceId(_instanceId).withSegmentName(_segmentNameStr).withOffset(_offset);
+    SegmentCompletionProtocol.Response response = SegmentCompletionManager.getInstance().segmentCommitEnd(reqParams, true, true);
```
Comment (mcvsubbu) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/LLCSegmentCommitEnd.java]: Before this line, add code to verify that the file exists.

File: pinot-common/src/main/java/com/linkedin/pinot/common/protocols/SegmentCompletionProtocol.java [hunk 11]
```diff
@@ -292,12 +309,19 @@ public Response(String jsonRespStr) {
       } else {
         _buildTimeSeconds = buildTimeObj;
       }
+
+      boolean isSplitCommit = false;
+      if (jsonObject.containsKey(COMMIT_TYPE_KEY)) {
```
Comment (mcvsubbu) [inline pinot-common/src/main/java/com/linkedin/pinot/common/protocols/SegmentCompletionProtocol.java]: need to get the boolean value of the definition

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/LLCSegmentCommitEnd.java [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+import org.restlet.representation.StringRepresentation;
+import org.restlet.resource.ServerResource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class LLCSegmentCommitEnd extends ServerResource {
+  private static Logger LOGGER = LoggerFactory.getLogger(LLCSegmentCommitEnd.class);
+
+  public LLCSegmentCommitEnd() throws IOException {
```
Comment (mcvsubbu) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/LLCSegmentCommitEnd.java]: we dont need the ctor

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/LLCSegmentCommitUpload.java [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+    if (params == null) {
+      return new StringRepresentation(SegmentCompletionProtocol.RESP_FAILED.toJsonString());
+    }
+
+    LOGGER.info(params.toString());
+
+    boolean success = uploadSegment(params.getInstanceId(), params.getSegmentName(), params.getOffset());
+    if (success) {
+      LOGGER.info("Uploaded segment successfully");
+      return new StringRepresentation(SegmentCompletionProtocol.ControllerResponseStatus.UPLOAD_SUCCESS.toString());
```
Comment (mcvsubbu) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/LLCSegmentCommitUpload.java]: Add a TODO to include the generated file name in the response to server

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/LLCSegmentCommitEnd.java [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+
+  }
+
+  @HttpVerb("get")
+  @Description("Segment commit end for a LLC segment using split commit")
+  @Summary("Segment commit end for a LLC segment using split commit")
+  @Paths({"/" + SegmentCompletionProtocol.MSG_TYPE_COMMIT_END})
+  public Representation get() {
+    SegmentCompletionProtocol.Request.Params requestParams = SegmentCompletionUtils.extractParams(getReference());
```
Comment (mcvsubbu) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/LLCSegmentCommitEnd.java]: Add a TODO to extract and possibly verify the file name in the request

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/LLCSegmentCommitUpload.java [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+      final File tableDir = new File(baseDataDir, rawTableName);
+      final File segmentFile = new File(tableDir, SegmentCompletionUtils.generateSegmentFileName(segmentNameStr, offset, instanceId));
+
+      // Always delete the segment file if it exists for split commit.
+      if (segmentFile.exists()) {
+        LOGGER.warn("Segment file {} exists. Replacing with upload from {}", segmentNameStr, instanceId);
+        FileUtils.deleteQuietly(segmentFile);
+      }
+      FileUtils.moveFile(dataFile, segmentFile);
+
```
Comment (mcvsubbu) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/LLCSegmentCommitUpload.java]: Log info with generated file name

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/LLCSegmentCommitUpload.java [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+        LOGGER.error("Segment not included in request. Instance {}, segment {}", instanceId, segmentNameStr);
+        return false;
+      }
+      // We will not check for quota here. Instead, committed segments will count towards the quota of a
+      // table
+      LLCSegmentName segmentName = new LLCSegmentName(segmentNameStr);
+
+      final String rawTableName = segmentName.getTableName();
+      final File tableDir = new File(baseDataDir, rawTableName);
+      final File segmentFile = new File(tableDir, SegmentCompletionUtils.generateSegmentFileName(segmentNameStr, offset, instanceId));
```
Comment (mcvsubbu) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/LLCSegmentCommitUpload.java]: pull out the code to genearte file name as:
generatedFileName  = SCU.generateFileName(..)

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/LLCSegmentStoppedConsuming.java [hunk 17]
```diff
@@ -29,10 +29,7 @@
 
 
 public class LLCSegmentStoppedConsuming extends PinotSegmentUploadRestletResource {
-  private static Logger LOGGER = LoggerFactory.getLogger(LLCSegmentCommit.class);
-  long _offset;
-  String _segmentNameStr;
-  String _instanceId;
+  private static Logger LOGGER = LoggerFactory.getLogger(LLCSegmentStoppedConsuming.class);
```
Comment (mcvsubbu) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/LLCSegmentStoppedConsuming.java]: good catch

**Comment by jenniferdai** [comment]
Tested this by sending GET/POST requests manually. Integration tests will be added in the next part.

**Pull request merged**